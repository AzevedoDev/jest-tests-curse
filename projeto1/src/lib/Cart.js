import { find, remove } from 'lodash';

export default function Cart() {
  let cart = [];

  const add = item => {
    const itemToFind = { product: item.product };
    if (find(cart, itemToFind)) {
      remove(cart, itemToFind);
    }
    cart.push(item);
  };

  const removeItem = product => {
    remove(cart, { product });
  };

  const getTotal = () =>
    cart.reduce((acc, item) => acc + item.quantity * item.product.price, 0);

  const resume = () => ({ total: getTotal(), items: cart });
  const checkout = () => {
    const { total, items } = resume();
    cart = [];
    return { total, items };
  };
  return {
    add,
    removeItem,
    getTotal,
    resume,
    checkout,
  };
}
