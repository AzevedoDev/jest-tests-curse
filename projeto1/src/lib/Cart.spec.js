import Cart from './Cart';

let cart;
const product = {
  title: 'Adidas running shoes - men',
  price: 35388,
};
const product2 = {
  title: 'Adidas running shoes - women',
  price: 41872,
};

beforeEach(() => {
  cart = new Cart();
});

describe('Cart', () => {
  describe('getTotal(', () => {
    it('should return 0 when getTotal() is executed in a newly created instance', () => {
      expect(cart.getTotal()).toEqual(0);
    });
    it('should multiply quantity and price and receive the total amount', () => {
      const item = {
        product,
        quantity: 2,
      };
      cart.add(item);
      expect(cart.getTotal()).toEqual(70776);
    });
    it('should ensure no more than on product exists at a time', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product,
        quantity: 1,
      });
      expect(cart.getTotal()).toEqual(35388);
    });
    it('should update total when a product gets included and then removed', () => {
      cart.add({ product, quantity: 2 });
      cart.add({ product: product2, quantity: 1 });
      cart.removeItem(product);

      expect(cart.getTotal()).toEqual(41872);
    });
  });
  describe('checkout()', () => {
    it('should return an object with total and list of items', () => {
      cart.add({
        product,
        quantity: 2,
      });
      cart.add({
        product: product2,
        quantity: 1,
      });
      expect(cart.checkout()).toMatchSnapshot();
    });
    it('should return an object with total and list of items when call resume()', () => {
      cart.add({
        product,
        quantity: 2,
      });
      cart.add({
        product: product2,
        quantity: 1,
      });
      expect(cart.resume()).toMatchSnapshot();
    });
    it('should reset cart when checkout is called', () => {
      cart.add({
        product,
        quantity: 2,
      });
      cart.add({
        product: product2,
        quantity: 1,
      });
      cart.checkout();
      expect(cart.getTotal()).toEqual(0);
    });
  });
});
