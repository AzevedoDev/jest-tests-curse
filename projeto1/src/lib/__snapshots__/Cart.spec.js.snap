// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout() should return an object with total and list of items 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Adidas running shoes - men",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 41872,
        "title": "Adidas running shoes - women",
      },
      "quantity": 1,
    },
  ],
  "total": 112648,
}
`;

exports[`Cart checkout() should return an object with total and list of items when call resume() 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Adidas running shoes - men",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 41872,
        "title": "Adidas running shoes - women",
      },
      "quantity": 1,
    },
  ],
  "total": 112648,
}
`;
